server/src/main/java/org/elasticsearch/tasks/CancellableTasksTracker.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+/**
+ * Tracks items that are associated with cancellable tasks, supporting efficient lookup by task ID and by parent task ID
+ */
+public class CancellableTasksTracker<T> {
+
+    private final T[] empty;
```

Comment:
Can `empty` be a constant? It doesn't seems to modified anywhere and it's used in `Arrays.copyOf`.

Reply:
Again I didn't find a way to do so, but you're welcome to show me how :)

Reply:
Why does this have to be a generic, `T == CancellableTaskHolder` is the only use case we have isn't it?

Reply:
> I don't think it makes a performance difference. I could be persuaded.

Nah there isn't any difference I think lets leave it as is then IMO :)